File: sketch/game/GameMenu/Inventory/Item.ts
class Item {
  private id: itemList;
  private InventoryPosX: number;
  private InventoryPosY: number;
  width: number = 64;
  img: p5.Image;
  private stackSize: number;
  private abilities: Abilities[];

  constructor(obj: itemInput) {
    this.InventoryPosX = obj.InventoryPosX;
    this.InventoryPosY = obj.InventoryPosY;
    this.id = obj.id;
    this.img = obj.img;
    this.stackSize = obj.stackSize;
    this.abilities = obj.abilities;
  }

  showItem() {
    let itemBoxOffset = (Inventory.SLOTSIZE - this.width) / 2;
    let offsetX = width / 2 - (Inventory.SLOTSIZE * Inventory.BACKPACKWITDH) / 2;
    let offsetY = height / 2 - (Inventory.SLOTSIZE * Inventory.BACKPACKHEIGHT) / 2;
    image(this.img, this.InventoryPosX * Inventory.SLOTSIZE + offsetX + itemBoxOffset, this.InventoryPosY * Inventory.SLOTSIZE + offsetY + itemBoxOffset, this.width, this.width);
  }
  // showItem() {
  //   push();

  //   let itemBoxOffset = (Inventory.SLOTSIZE - this.width) / 2;
  //   let offsetX = width / 2 - (Inventory.SLOTSIZE * Inventory.BACKPACKWITDH) / 2;
  //   let offsetY = height / 2 - (Inventory.SLOTSIZE * Inventory.BACKPACKHEIGHT) / 2;
  //   fill(255, 255, 255);
  //   rect(this.InventoryPosX * Inventory.SLOTSIZE + offsetX + itemBoxOffset, this.InventoryPosY * Inventory.SLOTSIZE + offsetY + itemBoxOffset, this.width, this.width);
  //   pop();
  // }

  getStackSize(): number {
    return this.stackSize;
  }

  getItem(): Item {
    return this;
  }

  getInventoryPos(): Coords {
    return { x: this.InventoryPosX, y: this.InventoryPosY };
  }

  getId(): number {
    return this.id;
  }

  itemSelected(): void {
    this.abilities.forEach((element) => {
      element.abilitySelected();
    });
  }

  clicked(playerX: number, playerY: number) {
    this.abilities.forEach((element) => {
      element.abilityClicked(playerX, playerY);
    });
  }

  tick(playerX: number, playerY: number) {
    this.abilities.forEach((element) => {
      element.abilityTick(playerX, playerY);
    });
  }
}

interface itemInput {
  InventoryPosX: number;
  InventoryPosY: number;
  id: itemList;
  stackSize: number;
  img: p5.Image;
  abilities: Abilities[];
}

File: sketch/game/GameMenu/Inventory/Item.java
package sketch.game.GameMenu.Inventory;

import java.util.ArrayList;

import sketch.game.utilities.Constants;
import sketch.game.utilities.Point;

public class Item {
  
  private Constants.itemList id;
  private int InventoryPosX;
  private int InventoryPosY;
  int width = 64;
  p5.Image img;
  private int stackSize;
  private ArrayList<Abilities> abilities;

  Item(itemInput obj) {
    this.InventoryPosX = obj.InventoryPosX;
    this.InventoryPosY = obj.InventoryPosY;
    this.id = obj.id;
    this.img = obj.img;
    this.stackSize = obj.stackSize;
    this.abilities = obj.abilities;
  }

  void showItem() {
    int itemBoxOffset = (Inventory.SlotSize - this.width) / 2;
    int offsetX = width / 2 - (Inventory.SlotSize * Inventory.BACKPACKWITDH) / 2;
    int offsetY = height / 2 - (Inventory.SlotSize * Inventory.BACKPACKHEIGHT) / 2;
    image(this.img, this.InventoryPosX * Inventory.SlotSize + offsetX + itemBoxOffset, this.InventoryPosY * Inventory.SlotSize + offsetY + itemBoxOffset, this.width, this.width);
  }

  public int getStackSize() {
    return this.stackSize;
  }

  public Item getItem() {
    return this;
  }

  public Point getInventoryPos() {
    return new Point(this.InventoryPosX, this.InventoryPosY);
  }

  Constants.itemList getId() {
    return this.id;
  }

  public void itemSelected() {
    for (int i = 0; i < this.abilities.size(); i++) {
      this.abilities.get[i].abilitySelected();
    }
  }

  public void clicked(int playerX, int playerY) {
    for (int i = 0; i < this.abilities.size(); i++) {
      this.abilities.get[i].abilityClicked(playerX, playerY);
    }
  }

  public void tick(int playerX,int playerY) {
    for (int i = 0; i < this.abilities.size(); i++) {
      this.abilities.get[i].abilityTick(playerX, playerY);
    }
  }
}

interface itemInput {
  int InventoryPosX;
  int InventoryPosY;
  Constants.itemList id;
  int stackSize;
  p5.Image img;
  ArrayList<Abilities> abilities;
}
